### PR TITLE
implement yolo v2 training

### PR DESCRIPTION
**Pull request by ryansun1900**

### CODE REVIEW

**Comment by thtrieu** [comment]
You did it @ryansun1900   :)
Just to make sure you did it correctly, can you verify that you are able to overfit a small training dataset (about ~10 images on various classes of object) to nearly perfect accuracy? Please do it, don't forget to disable the data augmentation when doing so.

I'll merge this request as soon as you were able to post the result here.
Cheers!

**Comment by ryansun1900** [comment]
1. put 10 images for training:
data [['2007_000068.jpg', [500, 375, [['bird', 27, 45, 266, 375]]]], ['2007_000027.jpg', [486, 500, [['person', 174, 101, 349, 351]]]], ['2007_000042.jpg', [500, 335, [['train', 263, 32, 500, 295], ['train', 1, 36, 235, 299]]]], ['2007_000063.jpg', [500, 375, [['dog', 123, 115, 379, 275], ['chair', 75, 1, 428, 375]]]], ['2007_000033.jpg', [500, 366, [['aeroplane', 9, 107, 499, 263], ['aeroplane', 421, 200, 482, 226], ['aeroplane', 325, 188, 411, 223]]]], ['2007_000123.jpg', [500, 375, [['train', 1, 26, 358, 340]]]], ['2007_000061.jpg', [500, 333, [['boat', 274, 11, 437, 279], ['boat', 184, 214, 281, 252]]]], ['2007_000039.jpg', [500, 375, [['tvmonitor', 156, 89, 344, 279]]]], ['2007_000032.jpg', [500, 281, [['aeroplane', 104, 78, 375, 183], ['aeroplane', 133, 88, 197, 123], ['person', 195, 180, 213, 229], ['person', 26, 189, 44, 238]]]], ['2007_000121.jpg', [500, 375, [['tvmonitor', 251, 28, 475, 267], ['tvmonitor', 22, 28, 251, 273]]]]]

2. disable the augmentation
in "def imcv2_affine_trans(im)"
set scale = 1 & flip = 0 directly.

3. set the learning rate
flags.DEFINE_float("lr", 1e-1, "learning rate") 

4. run the command:
./flow --model cfg/tiny-yolo-voc.cfg  --train

5. here is my result
Training statistics: 
	Learning rate : 0.1
	Batch size    : 10
	Epoch number  : 1000
	Backup every  : 2000
step 1 - loss 107.437637329 - moving ave loss 107.437637329
Finish 1 epoch(es)
step 2 - loss 53.9329414368 - moving ave loss 102.08716774
Finish 2 epoch(es)
step 3 - loss 5.46648311615 - moving ave loss 92.4250992775
Finish 3 epoch(es)
step 4 - loss 5.46170139313 - moving ave loss 83.7287594891
Finish 4 epoch(es)
step 5 - loss 5.45717334747 - moving ave loss 75.9016008749
Finish 5 epoch(es)


6.use the pre-trained weight "tiny-yolo-voc.weights"
command: 
./flow --model cfg/tiny-yolo-voc.cfg --load bin/tiny-yolo-voc.weights --train

here is the loss for this weight:
step 1 - loss 3.31776690483 - moving ave loss 3.31776690483
Finish 1 epoch(es)

**Comment by thtrieu** [comment]
Can you post some of the images with bounding box too ?

**Comment by ryansun1900** [comment]
Would you mean that the 10 images I used in training with true label bounding box?

data [['2007_000068.jpg', [500, 375, [['bird', 27, 45, 266, 375]]]], ['2007_000027.jpg', [486, 500, [['person', 174, 101, 349, 351]]]], ['2007_000042.jpg', [500, 335, [['train', 263, 32, 500, 295], ['train', 1, 36, 235, 299]]]], ['2007_000063.jpg', [500, 375, [['dog', 123, 115, 379, 275], ['chair', 75, 1, 428, 375]]]], ['2007_000033.jpg', [500, 366, [['aeroplane', 9, 107, 499, 263], ['aeroplane', 421, 200, 482, 226], ['aeroplane', 325, 188, 411, 223]]]], ['2007_000123.jpg', [500, 375, [['train', 1, 26, 358, 340]]]], ['2007_000061.jpg', [500, 333, [['boat', 274, 11, 437, 279], ['boat', 184, 214, 281, 252]]]], ['2007_000039.jpg', [500, 375, [['tvmonitor', 156, 89, 344, 279]]]], ['2007_000032.jpg', [500, 281, [['aeroplane', 104, 78, 375, 183], ['aeroplane', 133, 88, 197, 123], ['person', 195, 180, 213, 229], ['person', 26, 189, 44, 238]]]], ['2007_000121.jpg', [500, 375, [['tvmonitor', 251, 28, 475, 267], ['tvmonitor', 22, 28, 251, 273]]]]]

**Comment by ryansun1900** [comment]
I try to use one image to train (2007_000027.jpg)
set learning rate = 1e-2, 
command = "./flow --model cfg/tiny-yolo-voc.cfg  --train"

Result:
step 1 - loss 107.006942749 - moving ave loss 107.006942749
Finish 1 epoch(es)
...
step 56 - loss 0.0368079952896 - moving ave loss 1.27174379494
Finish 56 epoch(es)
step 57 - loss 0.0339030474424 - moving ave loss 1.14795972019
Finish 57 epoch(es)
step 58 - loss 0.0313070304692 - moving ave loss 1.03629445122
Finish 58 epoch(es)
step 59 - loss 0.0290170181543 - moving ave loss 0.935566707911
Finish 59 epoch(es)
step 60 - loss 0.0269409772009 - moving ave loss 0.84470413484
Finish 60 epoch(es)

**Comment by thtrieu** [comment]
I mean images with bounding boxes drawn onto it. You cannot tell the accuracy solely base on the loss, People want to look at the bounding boxes on the actual images too :)

**Comment by ryansun1900** [comment]
I want to make sure about after I training with 60 times. It will save some files in the ./ckpt/ folder such likes "tiny-yolo-voc-60.meta".
Then I just run the command : ./flow --model cfg/tiny-yolo-voc.cfg  --load 60.
It will use the weight I trained to make prediction?

**Comment by thtrieu** [comment]
Yes. That's the way it work :)

**Comment by ryansun1900** [comment]
I notice that it's hard to do perfect overfitting  in a small data for yolov2 training.

My guessing:
If we train with a very small data, all weights not in the last layer are still "random". We just tune the last layer's weight.
For yolov1, it could do perfect overfitting because  the last layer is "fully connected". We could tune the specific weight just for specific grid cell's class probability & bounding box's confidence in this grid cell.
But for yolov2, the last layer is convolution, if we just want to tune a specific bounding cox(in some grid cell)'s confidence & class probability, it will effect all the grid cells. Because all grids use the same filter!

So from my observation in doing perfect overfitting,  I get many bounding box in testing even I get a loss which is close to 0 in training.

**Comment by thtrieu** [comment]
Please confirm that the moving average (specifically moving variance) is
actually being updated for each training step, otherwise there is no chance
of overfitting anything.

**Comment by ryansun1900** [comment]
For the gamma/ moving_variance/ moving_mean, they are all the same value in each iteration. They are not be updated actually.

**Comment by lunardog** [comment]
Hi @ryansun1900, I think @thtrieu means to see that `moving ave loss` is updated at each step.

I'm trying to test this PR, but I can't get it to run. Things I tried:

1. in `framework.py`, lines 25 and 27 are commented, so the `YOLOv2` class still uses `vanilla` training, I uncommented
1. the new `train.py` and `data.py` are not included in `yolov2/__init__.py`, so they don't get imported in `feamework.py` even after uncommenting
1. syntax is python2, while the rest of the project is migrated to 3 
1. no `cPickle` in python 3 :(
1. some unused imports that break execution (`from misc import show` in `train.py`)

I'll keep trying, but are you sure you included all your changes in the pull request, @ryansun1900 ?

**Comment by ryansun1900** [comment]
We are discussing the batch norm layer actually. Because we found something wrong in the training for this layer.
I miss to pull some codes. After we fixed all problems and confirmed all things is OK. I will pull a new version. :)

**Comment by thtrieu** [comment]
Hey Ryan,

I fixed the batch-norm. You can now proceed to test your application.

Good luck.
